assignments/tdecacqu/notes-06.org

```diff
@@ -0,0 +1,320 @@
+#+TITLE: SICP 2.2 assignment
+#+AUTHOR: Tristan de Cacqueray
+#+BABEL: :cache yes
+#+PROPERTY: header-args :tangle yes
+
+* Building Abstractions with Data
+** Hierarchical Data and the Closure Property
+
+*** Exercise 2.17
+
+#+BEGIN_SRC scheme :tangle 2.17.scm
+  (define (last-pair l)
+    (if (null? (cdr l))
+        l
+        (last-pair (cdr l))))
+#+END_SRC
+
+*** Exercise 2.18
+
+#+BEGIN_SRC scheme :tangle 2.18.scm
+  (define (reverse l)
+    (define (iter l r)
+      (if (null? l)
+          r
+          (iter (cdr l) (cons (car l) r))))
+    (iter l '()))
+#+END_SRC
+
+*** Exercise 2.19
+
+Given
+
+#+BEGIN_SRC scheme :tangle 2.19.scm
+  (define us-coins (list 50 25 10 5 1))
+  (define uk-coins (list 100 50 20 10 5 2 1 0.5))
+  (define (cc amount coin-values)
+    (cond ((= amount 0) 1)
+          ((or (< amount 0) (no-more? coin-values)) 0)
+          (else
+           (+ (cc amount
+                  (except-first-denomination coin-values))
+              (cc (- amount
+                     (first-denomination coin-values))
+                  coin-values)))))
+#+END_SRC
+
+Define the procedures first-denomination, except-first-denomination, and no-more? in terms of
+primitive operations on list structures.
+
+#+BEGIN_SRC scheme :tangle 2.19.scm
+  (define (first-denomination l)
+    (car l))
+  (define (except-first-denomination l)
+    (cdr l))
+  (define (no-more? l)
+    (null? l))
+#+END_SRC
+
+The order doesn't affect the answer.
+
+*** Exercise 2.20
+
+Use the dotted-tail notation to write a procedure same-parity.
+
+#+BEGIN_SRC scheme :tangle 2.20.scm
+  (define (same-parity . l)
```

Comment:
This is a bit wrong actually, the original exercise requires "one or more". But this allows zero parameters also:

```scheme
scheme@(guile-user) [1]> (define (same-parity . y) y)
scheme@(guile-user) [1]> (same-parity 1)
$3 = (1)
scheme@(guile-user) [1]> (same-parity)
$4 = ()
scheme@(guile-user) [1]> 
```

Where as ``x . y`` requires the first parameter (that is, one or more parameters):

```scheme
scheme@(guile-user)> (define (same-parity x . y) y)
scheme@(guile-user)> (same-parity 1)
$2 = ()
scheme@(guile-user)> (same-parity)
;;; <stdin>:28:0: warning: possibly wrong number of arguments to `same-parity'
<unnamed port>:24:0: In procedure same-parity:
Wrong number of arguments to #<procedure same-parity (x . y)>

Entering a new prompt.  Type `,bt' for a backtrace or `,q' to continue.
scheme@(guile-user) [1]>
```

Reply:
I see thanks, this actually makes the procedure simpler.